### PR TITLE
Add a common failover handling for tikv and tiflash failover

### PR DESCRIPTION
**Pull request by lalitkfk**
### What problem does this PR solve?
The code handling the failover of TiFlash and TiKV are quite similar but are present in their respective handlers which can be refactored to a common logic.

### What is changed and how does it work?
Move the common logic of handling failover in both tiflash and tikv to a common store-failover and create an abstraction which can work on TidbCluster object to access the properties of tiflash and tikv.

Create a `commonStoreFailover` that handles the common failover logic for TiKV and TiFlash. Abstract out the part which has to access the TiKV or TiFlash specific properties into a `StoreAccess` interface. 

1. `tiflashStoreAccess` implements `StoreAccess` as a folder of access functions for TiFlash properties.
2. `tikvStoreAccess` implements `StoreAccess` as a folder of access functions for TiKV properties.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests
- [x] Unit test
- [x] E2E test
- [x] Manual test
- - Deploy the tidb operator
- - Create a tidb cluster and simulate pod failures for tikv and tiflash
- - Observe that the FailureStores are added in tikv and tiflash
- - Recover the tikv failed pod and observe that the FailureStores and FailoverUID are cleared in Tikv status
- - Recover the tiflash failed pod and observe that the FailureStores and FailoverUID are cleared in Tiflash status
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch
- [ ] Need to update the documentation

### Release Notes
Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

**Comment by lalitkfk** [comment]
Hi @hanlins 
Could you please take a look. I can refactor [PR](https://github.com/pingcap/tidb-operator/pull/4788) along the same lines as well once this is done.

**Comment by lalitkfk** [comment]
Hi @hanlins 
Could you please take a look.

File: pkg/manager/member/shared_store_failover.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+			// otherwise an extra replacement pod will be created.
+			delete(ssf.storeAccess.GetFailureStores(tc), key)
+		}
+	}
+}
+
+func (ssf *sharedStoreFailover) Recover(tc *v1alpha1.TidbCluster) {
+	ssf.storeAccess.ClearFailStatus(tc)
+	klog.Infof("%s recover: clear FailureStores, %s/%s", ssf.storeAccess.GetMemberType(), tc.GetNamespace(), tc.GetName())
+}
```
Comment (hanlins) [inline pkg/manager/member/shared_store_failover.go:129]: Can we add
```suggestion
}

var _ Failover = (*sharedStoreFailover)(nil)
```
to enforce compiler check that `sharedStoreFailover` is an implementation of `Failover`? Might also be useful if we can have a `NewFailover` which takes `StoreCommonAccess` as input and returns a `Failover` instance (similar to `NewTiKVFailover`).

File: pkg/manager/member/tiflash_failover.go [hunk 2]
```diff
@@ -14,104 +14,75 @@
...
-	"github.com/pingcap/tidb-operator/pkg/util"
-	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/uuid"
-	"k8s.io/klog/v2"
 )
 
-// TODO reuse tikvFailover since we share the same logic
 type tiflashFailover struct {
```
Comment (hanlins) [inline pkg/manager/member/tiflash_failover.go:23]: Is it possible if we eliminate `tiflashFailover` and use `sharedStoreFailover` instead? Maybe just keep the implementation of `tiflashStoreAccess` in this file?

File: pkg/manager/member/tikv_failover.go [hunk 3]
```diff
@@ -14,104 +14,75 @@
...
 	"github.com/pingcap/tidb-operator/pkg/controller"
-	"github.com/pingcap/tidb-operator/pkg/util"
-	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/uuid"
-	"k8s.io/klog/v2"
 )
 
 type tikvFailover struct {
```
Comment (hanlins) [inline pkg/manager/member/tikv_failover.go:23]: Similar suggestion to TiFlash above.

File: pkg/manager/member/shared_store_failover.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	GetStsDesiredOrdinals(tc *v1alpha1.TidbCluster, excludeFailover bool) sets.Int32
+}
+
+// sharedStoreFailover contains the shared failover logic of TiKV and TiFlash
+type sharedStoreFailover struct {
+	storeAccess StoreCommonAccess
+	deps        *controller.Dependencies
+}
+
+func (ssf *sharedStoreFailover) doFailover(tc *v1alpha1.TidbCluster, failoverPeriod time.Duration) error {
```
Comment (hanlins) [inline pkg/manager/member/shared_store_failover.go:48]: Maybe just make rename it as `Failover` to make `sharedStoreFailover` an implementation of `Failover`.

**Review by hanlins — COMMENTED** [review]
Hi @lalitkfk, thanks for the great contributions unifying failure handling for TiFlash and TiKV!
It's great that we've introduced a common interface `StoreCommonAccess` which abstracts the access for both TiKV and TiFlash, and leveraging that interface, we've created `sharedStoreFailover` which is effectively an implementation for `Failover`.
Have some suggestions to further improve this changeset:
1. Maybe we can replace TiKV and TiFlash's own `Failover` implementation (i.e. `tikvFailover` and `tiflashFailover` respectively) with `sharedStoreFailover` directly. Adding a wrapper on top of `sharedStoreFailover` seems unnecessary.
2. In controller, we can create `sharedStoreFailover` (maybe via a new method `NewFailover` that takes TiKV/TiFlash's implementation of `StoreCommonAccess` as an input) directly instead of creating `tiflashFailover` which wraps `sharedStoreFailover`. Here's for TiKV:
https://github.com/pingcap/tidb-operator/blob/35f941e65a272db837d3e44e4d1a9679e0917698/pkg/controller/tidbcluster/tidb_cluster_controller.go#L58
Here's TiFlash:
https://github.com/pingcap/tidb-operator/blob/35f941e65a272db837d3e44e4d1a9679e0917698/pkg/controller/tidbcluster/tidb_cluster_controller.go#L66

Let me know if you have other concerns.

**Comment by lalitkfk** [comment]
Thanks for the review. It is actually a cool suggestion. I will try to do that. I am also thinking to rename sharedStoreFailover to commonStoreFailover to denote that it is a common Failover handler for TiKV and TiFlash. The failoverPeriod will be different for TiKV and TiFlash which comes from the operator's CLIConfig, for which I am thinking to have failoverPeriod as a field in the commonStoreFailover which will be initialised accordingly in the NewTiFlashFailover or NewTiKVFailover methods.

**Comment by hanlins** [comment]
> Thanks for the review @hanlins . It is actually a cool suggestion. I will try to do that. I am also thinking to rename sharedStoreFailover to commonStoreFailover to denote that it is a common Failover handler for TiKV and TiFlash. The failoverPeriod will be different for TiKV and TiFlash which comes from the operator's CLIConfig, for which I am thinking to have failoverPeriod as a field in the commonStoreFailover which will be initialised accordingly in the NewTiFlashFailover or NewTiKVFailover methods.

That would be great! The renamed `commonStoreFailover`  makes more sense to me 👍. If necessary, we could add `failoverPeriod` to the `StoreCommonAccess` interface, if the value differed for TiFlash and TiKV.

**Comment by lalitkfk** [comment]
Hi @hanlins I have addressed the comments. Also updated the PR description. Please take a look.

File: pkg/manager/member/common_store_failover.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+func (fsf *fakeStoreFailover) Failover(_ *v1alpha1.TidbCluster) error {
+	return nil
+}
+
+func (fsf *fakeStoreFailover) Recover(_ *v1alpha1.TidbCluster) {
+}
+
+func (fsf *fakeStoreFailover) RemoveUndesiredFailures(_ *v1alpha1.TidbCluster) {
+}
```
Comment (hanlins) [inline pkg/manager/member/common_store_failover.go:142]:
```suggestion
func (fsf *fakeStoreFailover) Recover(_ *v1alpha1.TidbCluster) {}

func (fsf *fakeStoreFailover) RemoveUndesiredFailures(_ *v1alpha1.TidbCluster) {}

var _ Failover = (*fakeStoreFailover)(nil)
```

File: pkg/manager/member/common_store_failover.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	ClearFailStatus(tc *v1alpha1.TidbCluster)
+	GetStsDesiredOrdinals(tc *v1alpha1.TidbCluster, excludeFailover bool) sets.Int32
+}
+
+// commonStoreFailover handles the common failover logic for TiKV and TiFlash
+type commonStoreFailover struct {
+	deps        *controller.Dependencies
+	storeAccess StoreAccess
+}
+
```
Comment (hanlins) [inline pkg/manager/member/common_store_failover.go:48]: Explicitly enforce static check that struct `commonStoreFailover` implements `Failover` interface.
```suggestion

var _ Failover = (*commonStoreFailover)(nil)

```

**Comment by lalitkfk** [comment]
Thanks for the review @hanlins. I have made the suggested changes.

**Review by hanlins — APPROVED** [review]
Thanks for the contribution @lalitkfk, LGTM!

**Comment by lalitkfk** [comment]
Hi @hanlins How do I take this forward and get it merged with master branch ?

**Comment by hanlins** [comment]
> Hi @hanlins How do I take this forward and get it merged with master branch ?

Hi @lalitkfk, nothing required from your end, it needs another approval to be merged automatically. I've requested @csuzhangxc to take a look on this.

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**